examples/src/rescue_raps/mod.rs

```diff
@@ -0,0 +1,172 @@
+        assert!(
+            chain_length.is_power_of_two(),
+            "chain length must a power of 2"
+        );
+
+        let mut seeds = vec![[BaseElement::ZERO; 2]; chain_length as usize];
+        for internal_seed in seeds.iter_mut() {
+            *internal_seed = rand_array();
+        }
+        let mut permuted_seeds = seeds[2..].to_vec();
+        permuted_seeds.push(seeds[0]);
+        permuted_seeds.push(seeds[1]);
+
+        // compute the sequence of hashes using external implementation of Rescue hash
+        let now = Instant::now();
+        let result = compute_permuted_hash_chains(&seeds, &permuted_seeds);
+        debug!(
+            "Computed two permuted chains of {} Rescue hashes in {} ms",
+            chain_length,
+            now.elapsed().as_millis(),
+        );
+
+        RescueRapsExample {
+            options,
+            chain_length,
+            seeds,
+            permuted_seeds,
+            result,
+        }
+    }
+}
+
+// EXAMPLE IMPLEMENTATION
+// ================================================================================================
+
+impl Example for RescueRapsExample {
+    fn prove(&self) -> StarkProof {
+        // generate the execution trace
+        debug!(
+            "Generating proof for computing a chain of {} Rescue hashes\n\
+            ---------------------",
+            self.chain_length
+        );
+
+        // create a prover
+        let prover = RescueRapsProver::new(self.options.clone());
+
+        // generate the execution trace
+        let now = Instant::now();
+        let trace = prover.build_trace(&self.seeds, &self.permuted_seeds, self.result);
+        let trace_length = trace.length();
+        debug!(
+            "Generated execution trace of {} registers and 2^{} steps in {} ms",
+            trace.width(),
+            log2(trace_length),
+            now.elapsed().as_millis()
+        );
+
+        // generate the proof
+        prover.prove(trace).unwrap()
+    }
+
+    fn verify(&self, proof: StarkProof) -> Result<(), VerifierError> {
+        let pub_inputs = PublicInputs {
+            result: self.result,
+        };
+        winterfell::verify::<RescueRapsAir>(proof, pub_inputs)
+    }
+
+    fn verify_with_wrong_inputs(&self, proof: StarkProof) -> Result<(), VerifierError> {
+        let pub_inputs = PublicInputs {
+            result: [self.result[1], self.result[0]],
+        };
+        winterfell::verify::<RescueRapsAir>(proof, pub_inputs)
+    }
+}
+
+// HELPER FUNCTIONS
+// ================================================================================================
+
+fn merge(values: &[[BaseElement; 2]; 2], result: &mut [BaseElement]) {
+    let mut state = BaseElement::zeroed_vector(STATE_WIDTH);
+    state[..2].copy_from_slice(&values[0]);
+    state[2..].copy_from_slice(&values[1]);
+    for i in 0..NUM_HASH_ROUNDS {
+        rescue::apply_round(&mut state, i);
+    }
+    result.copy_from_slice(&state[..2]);
+}
```

Comment:
I think the naming is a bit confusing here. Usually, when we say `merge` we mean to merge two independent values, in which case, the capacity elements of the state are initialized to zeros.

Probably a more appropriate name would be `absorb`. In such cases we take the previous state and add values to be absorbed into the rate portion of the state. It could look something like this:

```Rust
fn absorb(state: &mut [BaseElement; STATE_WIDTH], values: [BaseElement; 2]) {
    state[0] += values[0];
    state[1] += values[1];
    for i in 0..NUM_HASH_ROUNDS {
        rescue::apply_round(&mut state, i);
    }
}
```

It doesn't make a big difference for example purposes, but if possible - would be great to update it.

Reply:
Ah yes, sorry for the naming. I changed the example. This requires one extra periodic column, to retrieve the inserted values as we don't reset registers but add them instead, though I believe this is still fine.